controllers/cohortdata.go

```diff
@@ -0,0 +1,157 @@
+		c.Abort()
+		return
+	}
+	decoder := json.NewDecoder(c.Request.Body)
+	var conceptIds ConceptIds
+	err := decoder.Decode(&conceptIds)
+	if err != nil {
+		log.Printf("Error: %s", err)
+		c.JSON(http.StatusBadRequest, gin.H{"message": "bad request - no request body"})
+		c.Abort()
+		return
+	}
+	log.Printf("Querying concept ids: %v", conceptIds.ConceptIds)
+
+	sourceId, _ := strconv.Atoi(sourceIdStr)
+	cohortId, _ := strconv.Atoi(cohortIdStr)
+
+	// call model method:
+	cohortData, err := cohortDataModel.RetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPersonId(sourceId, cohortId, conceptIds.ConceptIds)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error retrieving concept details", "error": err})
+		c.Abort()
+		return
+	}
+	b := GenerateTSV(sourceId, cohortData, conceptIds.ConceptIds)
+	c.String(http.StatusOK, b.String())
+	return
+}
+
+// This function will take the given cohort data and transform it into a matrix
+// that contains the person id as the first column and the concept values found
+// for this person in the subsequent columns. The transformation is necessary
+// since the cohortData list contains one row per person-concept combination.
+// E.g. the following (simplified version of the) data:
+//   {PersonId:1, ConceptId:1, ConceptValue: A},
+//   {PersonId:1, ConceptId:2, ConceptValue: B},
+//   {PersonId:2, ConceptId:1, ConceptValue: C},
+// will be transformed to a TSV table like:
+//   sample.id       concept_name_for_id1      concept_name_for_id2
+//   1               A                         B
+//   2               C                         NA
+// where "NA" means that the person did not have a data element for that concept
+// or that the data element had a NULL/empty value.
+func GenerateTSV(sourceId int, cohortData []*models.PersonConceptAndValue, conceptIds []int) *bytes.Buffer {
+	b := new(bytes.Buffer)
+	w := csv.NewWriter(b)
+	w.Comma = '\t'
+
+	var rows [][]string
+	var header []string
+	header = append(header, "sample.id")
+	header = addConceptsToHeader(sourceId, header, conceptIds)
+	rows = append(rows, header)
+
+	var currentPersonId = -1
+	var row []string
+	for _, cohortDatum := range cohortData {
+		// if new person, start new row:
+		if cohortDatum.PersonId != currentPersonId {
+			if currentPersonId != -1 {
+				rows = append(rows, row)
+			}
+			row = []string{}
+			row = append(row, strconv.Itoa(cohortDatum.PersonId))
+			row = appendInitEmptyConceptValues(row, len(conceptIds))
+			currentPersonId = cohortDatum.PersonId
+		}
+		row = populateConceptValue(row, *cohortDatum, conceptIds)
+	}
+	// append last person row:
+	rows = append(rows, row)
+
+	// TODO - is there a way to write as the rows are produced? Building up all rows in memory
+	// could cause issues if the cohort vs concepts matrix gets very large...or will the number of concepts
+	// queried at the same time never be very large? Should we restrict the number of concepts to
+	// a max here in this method?
```

Comment:
question for @kmhernan  ^

Reply:
In gin streaming can be done with [Stream](https://pkg.go.dev/github.com/gin-gonic/gin#Context.Stream) function, for example [here](https://github.com/gin-gonic/examples/blob/6c5e6ca7d974e2dfe147da820a4101a7fd6ec9f8/realtime-chat/main.go#L32-L40).

Though, this is something we can wait on implementing, as it's not the highest priority (let's make it work first).